src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java

```diff
@@ -0,0 +1,289 @@
+		this.sequenceNumberExpression = sequenceNumberExpression;
+	}
+
+	/**
+	 * Specify a {@link OutboundMessageMapper} for embedding message headers into the record data
+	 * together with payload.
+	 * @param embeddedHeadersMapper the {@link OutboundMessageMapper} to embed headers into the record data.
+	 * @since 2.0
+	 * @see org.springframework.integration.support.json.EmbeddedJsonHeadersMessageMapper
+	 */
+	public void setEmbeddedHeadersMapper(OutboundMessageMapper<byte[]> embeddedHeadersMapper) {
+		this.embeddedHeadersMapper = embeddedHeadersMapper;
+	}
+
+	/**
+	 * Unsupported operation. Use {@link #setEmbeddedHeadersMapper} instead.
+	 * @param headerMapper is not used.
+	 * @see #setEmbeddedHeadersMapper
+	 */
+	@Override
+	public void setHeaderMapper(HeaderMapper<Void> headerMapper) {
+		throw new UnsupportedOperationException("Kinesis doesn't support headers.\n" +
+				"Consider to use 'OutboundMessageMapper<byte[]>' for embedding headers into the record data.");
+	}
+
+	@Override
+	protected Future<?> handleMessageToAws(Message<?> message) throws Exception {
+		if (message.getPayload() instanceof PutRecordsRequest) {
+			throw new UnsupportedOperationException("not implemented");
+		}
+		else {
+			final PutRecordRequest putRecordRequest =
+					(message.getPayload() instanceof PutRecordRequest)
+							? (PutRecordRequest) message.getPayload()
+							: buildPutRecordRequest(message);
+
+			// convert the PutRecordRequest to a UserRecord
+			UserRecord userRecord = new UserRecord();
+			userRecord.setExplicitHashKey(putRecordRequest.getExplicitHashKey());
+			userRecord.setData(putRecordRequest.getData());
+			userRecord.setPartitionKey(putRecordRequest.getPartitionKey());
+			userRecord.setStreamName(putRecordRequest.getStreamName());
+			return handleUserRecord(message, putRecordRequest, userRecord);
+		}
+	}
+
+	private Future<?> handleUserRecord(Message<?> message, final PutRecordRequest putRecordRequest,
+			UserRecord userRecord) {
+		ListenableFuture<UserRecordResult> recordResult = this.kinesisProducer.addUserRecord(userRecord);
+
+		final AsyncHandler<PutRecordRequest, PutRecordResult> asyncHandler =
+				obtainAsyncHandler(message, putRecordRequest);
+		final FutureCallback<UserRecordResult> callback = new FutureCallback<UserRecordResult>() {
+			@Override
+			public void onFailure(Throwable t) {
+				asyncHandler.onError((t instanceof Exception) ? ((Exception) t) : new AwsRequestFailureException(message, putRecordRequest, t));
+			}
+
+			@Override
+			public void onSuccess(UserRecordResult result) {
+				PutRecordResult putRecordResult = new PutRecordResult();
+				putRecordResult.setSequenceNumber(result.getSequenceNumber());
+				putRecordResult.setShardId(result.getShardId());
+				asyncHandler.onSuccess(putRecordRequest, putRecordResult);
+			}
+		};
+		Futures.addCallback(recordResult, callback, MoreExecutors.directExecutor());
+
+		return new FutureConverter<>(recordResult, userRecordResult -> {
```

Comment:
Since we are not going to convert a `UserRecordResult` into the `PutRecordResult`, I think we don't need this extra `FutureConverter` class at all and we can just call `asyncHandler` directly.